### PR TITLE
fix(walredo spawn): coalescing stalls other executors std::sync::RwLock

### PR DESCRIPTION
**Pull request by problame**
part of #6628

Before this PR, we used a std::sync::RwLock to coalesce multiple
callers on one walredo spawning. One thread would win the write lock
and others would queue up either at the read() or write() lock call.

In a scenario where a compute initiates multiple getpage requests
from different Postgres backends (= different page_service conns),
and we don't have a walredo process around, this means all these
page_service handler tasks will enter the spawning code path,
one of them will do the spawning, and the others will stall their
respective executor thread because they do a blocking
read()/write() lock call.

I don't know exactly how bad the impact is in reality because
posix_spawn uses CLONE_VFORK under the hood, which means that the
entire parent process stalls anyway until the child does `exec`,
which in turn resumes the parent.

But, anyway, we won't know until we fix this issue.
And, there's definitely a future way out of stalling the
pageserver on posix_spawn, namely, forking template walredo processes
that fork again when they need to be per-tenant.
This idea is tracked in https://github.com/neondatabase/neon/issues/7320.

Changes
-------

This PR fixes that scenario by switching to use `heavier_once_cell`
for coalescing. There is a comment on the struct field that explains
it in a bit more nuance.

### Alternative Design

An alternative would be to use tokio::sync::RwLock.
I did this in the first commit in this PR branch,
before switching to `heavier_once_cell`.

Performance
-----------

I re-ran the `bench_walredo` and updated the results, showing that
the changes are neglible.

For the record, the earlier commit in this PR branch that uses
`tokio::sync::RwLock` also has updated benchmark numbers, and the
results / kinds of tiny regression were equivalent to `heavier_once_cell`.

Note that the above doesn't measure performance on the cold path, i.e.,
when we need to launch the process and coalesce. We don't have a benchmark
for that, and I don't expect any significant changes. We have metrics
and we log spawn latency, so, we can monitor it in staging & prod.

Risks
-----

As "usual", replacing a std::sync primitive with something that yields to
the executor risks exposing concurrency that was previously implicitly
limited to the number of executor threads.

This would be the first one for walredo.

The risk is that we get descheduled while the reconstruct data is already there.
That could pile up reconstruct data.

In practice, I think the risk is low because once we get scheduled again, we'll
likely have a walredo process ready, and there is no further await point
until walredo is complete and the reconstruct data has been dropped.

This will change with async walredo PR #6548, and I'm well aware of it
in that PR.

### CODE REVIEW

File: pageserver/benches/bench_walredo.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@
 //!
 //! # Reference Numbers
 //!
-//! 2024-03-20 on i3en.3xlarge
+//! 2024-03-20 on i3en.3xlarge TODO re-run
```
Comment (problame) [inline pageserver/benches/bench_walredo.rs:30]: TODO

File: pageserver/src/walredo.rs [hunk 2]
```diff
@@ -152,30 +160,46 @@ impl PostgresRedoManager {
...
     }
 
     /// This type doesn't have its own background task to check for idleness: we
     /// rely on our owner calling this function periodically in its own housekeeping
     /// loops.
-    pub(crate) fn maybe_quiesce(&self, idle_timeout: Duration) {
-        if let Ok(g) = self.last_redo_at.try_lock() {
+    pub(crate) async fn maybe_quiesce(&self, idle_timeout: Duration) {
+        // awkward control flow because rustc isn't smart enough to detect that we don't
+        // hold the std::sync lock guard across the await point
```
Comment (koivunej) [inline pageserver/src/walredo.rs:172]: std::sync mention is probably out of date now

Comment (problame) [reply pageserver/src/walredo.rs]: No, that was referring to the last_redo_at lock guard, it was accurate (I added it in this PR).

The switch to heavier_once_cell makes it unnecessary.

File: pageserver/src/walredo.rs [hunk 3]
```diff
@@ -275,17 +281,14 @@ impl PostgresRedoManager {
                 // Avoid concurrent callers hitting the same issue.
                 // We can't prevent it from happening because we want to enable parallelism.
                 {
-                    let mut guard = self.redo_process.write().unwrap();
-                    match &*guard {
-                        Some(current_field_value) => {
-                            if Arc::ptr_eq(current_field_value, &proc) {
+                    match self.redo_process.get() {
+                        None => (),
```
Comment (koivunej) [inline pageserver/src/walredo.rs:285]: I would prefer to keep the enum order for these match arms.

I was also wondering is the None branch here ok.

I think it is correct:

1. first one observing an error will get true for `Arc::ptr_eq`, take the permit and hold it while doing `kill; wait;`
2. while the permit is being held, `get` will return `None`, so the rest of "error observers" will continue fast from here
3. until the deinit is ready, all retryers will wait to get the semaphore

Yes I think it works as it should here.

Comment (koivunej) [reply pageserver/src/walredo.rs:285]: Actually, there might be an implicit drop-order thing going on.

`Guard::take_and_deinit` returns `(T, InitPermit)`. I think that value is dropped like:

1. `T`
2. `InitPermit`

However, we could make this explicit by:

```patch
-guard.take_and_deinit();
+let (proc, permit) = guard.take_and_deinit();
+drop(proc);
+drop(permit);
```

This is exactly what the implicit version is already doing per my understanding ([reference](https://doc.rust-lang.org/reference/destructors.html)):
> The fields of a [tuple](https://doc.rust-lang.org/reference/types/tuple.html) are dropped in order.

But here it really matters to us that we hold the permit over the `wait+kill`, correct?

Comment (koivunej) [reply pageserver/src/walredo.rs:285]: Aah I missed there was the line L303R306 with `drop(proc)`.

Ok, so I had misunderstood: because `proc` is an Arc, we cannot guarantee that a drop is happening while we are holding the init permit, and perhaps that was not the aim here at all, I was just imaging it was?

File: Cargo.toml [hunk 4]
```diff
@@ -170,6 +170,7 @@ tokio-postgres-rustls = "0.11.0"
 tokio-rustls = "0.25"
 tokio-stream = "0.1"
 tokio-tar = "0.3"
+tokio-test = "0.4.3"
```
Comment (koivunej) [inline Cargo.toml:173]: Why was this added?

File: pageserver/src/walredo.rs [hunk 5]
```diff
@@ -121,10 +135,11 @@ impl PostgresRedoManager {
                 self.conf.wal_redo_timeout,
                 pg_version,
             )
+            .await
         }
     }
 
-    pub(crate) fn status(&self) -> Option<WalRedoManagerStatus> {
+    pub(crate) async fn status(&self) -> Option<WalRedoManagerStatus> {
```
Comment (koivunej) [inline pageserver/src/walredo.rs:142]: No longer async.

File: pageserver/src/walredo.rs [hunk 6]
```diff
@@ -191,42 +208,31 @@ impl PostgresRedoManager {
...
-                                *proc_guard = Some(Arc::clone(&proc));
-                                proc
-                            }
-                            Some(proc) => Arc::clone(proc),
-                        }
+            let proc: Arc<process::WalRedoProcess> =
+                match self.redo_process.get_or_init_detached().await {
+                    Ok(guard) => Arc::clone(&guard),
+                    Err(permit) => {
+                        // don't hold poison_guard, the launch code can bail
```
Comment (koivunej) [inline pageserver/src/walredo.rs:215]: I don't see the poison guard code anymore...

I am unsure why this does not need any poison support.

File: pageserver/src/walredo.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/walredo.rs:285]: > Ok, so I had misunderstood: because proc is an Arc, we cannot guarantee that a drop is happening while we are holding the init permit,

Exactly, this should be evident from the block comment starting in line 294 as well as the comment on field `redo_process`; but apparently it's only evident for me.

I'll try to rearrange the code a bit to make it more clear

File: pageserver/src/walredo.rs (hunk 6 above)
Comment (problame) [reply pageserver/src/walredo.rs:215]: We probably woudn't have needed the poison guard code in the rwlock version of the patch either.

I don't see why we'd need it here; the all-or-nothing-ness of a cell means we can't leave something in half-broken state.

File: Cargo.toml (hunk 4 above)
Comment (problame) [reply Cargo.toml:173]: It was from the now-removed Poison

File: pageserver/src/walredo.rs (hunk 3 above)
Comment (koivunej) [reply pageserver/src/walredo.rs]: ok yes I was partially remembering correctly. I didn't remember we wanted to accept N=async threads walredo processes.

well, it doesn't really matter, occurrences are so rare.

File: pageserver/src/walredo.rs [hunk 7]
```diff
@@ -53,7 +54,19 @@ pub struct PostgresRedoManager {
...
+    /// requests don't use the [`heavier_once_cell::Guard`] to keep ahold of the
+    /// their process object; we use [`Arc::clone`] for that.
+    /// This is primarily because earlier implementations that didn't  use [`heavier_once_cell`]
+    /// had that behavior; it's probably unnecessary.
+    /// The only merit of it is that if one walredo process encounters an error,
+    /// it can take it out of rotation (= using [`heavier_once_cell::Guard::take_and_deinit`].
+    /// and retry redo, thereby starting the new process, while other redo tasks might
+    /// still be using the old redo process. But, those other tasks will most likely
+    /// encounter an error as well, and errors are an unexpected condition anyway.
+    /// So, probably we could get rid of the `Arc` in the future.
```
Comment (koivunej) [inline pageserver/src/walredo.rs:68]: I think this part here will need the timeout, you haven't yet removed it.

File: pageserver/src/walredo.rs (hunk 6 above)
Comment (koivunej) [reply pageserver/src/walredo.rs:215]: Confirmed, no poison needed because the communication is still blocking.

**Review by koivunej — APPROVED** [review]
